design/open/spark/grand-theft-s3-client.md

```diff
@@ -0,0 +1,141 @@
+      available for users to use to access lakeFS metadata directly.
+	</td><td>
+	  Accesses stored metadata directly on S3.
+	  <td>
+  </tr><tr>
+	<td>
+	  lakeFSFS
+	</td>
+	<td>
+	  Reading and writing directly on lakeFS.
+	</td>
+	<td>
+	  Read ETags of uploaded objects to put them in lakeFS metadata.  In
+      _some_ Hadoop versions, the S3AFileSystem returns FileStatus objects
+      with a S3AFileStatus.getETag method.  Otherwise a separate call to S3
+      is needed.
+	</td>
+  </tr>
+</table>
+
+![David Niven as Sir Charles Litton, The Pink Panther][pink-panther-img]
+
+These Spark clients cannot work without a working S3 client[^1].  This is:
+
+* **Different** between our two clients.
+
+  The Spark metadata client supports _only_ authentication to S3 using
+  access keys or STS, lakeFSFS supports _only_ taking ("_stealing_") clients
+  from S3AFileSystem.
+
+* **Brittle**.
+
+  Some users cannot use the authentication methods that we make available to
+  them.  The thievery code in lakeFSFS is subtle and greatly depends on an
+  assumed underlying implementation; it can break when DataBricks introduce
+  new features.
+* **Uninformative** in the case of system or user error.
+
+  Users receive very poor error reports.  If they get as far as an S3 client
+  but it is misconfigured, S3 happily generates "400 Bad Request" messages.
+  If client theft fails, it generates a report of the _last_ failure --
+  probably not the most _important_ failure.
+
+There are numerous bug reports and user questions about this area.
+
+## What?
+
+We propose to:
+
+1. **Reduce friction.**  When S3A already works on a Spark installation,
+   users should typically not have to add _any_ S3-related configuration in
+   order to use lakeFS clients.
+1. **Unify** S3 client acquisition between the two clients.  Both clients will
+   support the same configuration options: clients "stolen" from the
+   underlying S3AFileSystem, and explicitly created clients with static
+   access keys and STS.  Prefer stealing clients to creating them -- these
+   are most likely to work.
+1. **Improve** error reporting.  Report the stages attempted and how each
+   one failed.
+1. **Create a more general scheme** for generating clients.  Over time we
+   can hope to support more underlying implementations.
+
+## Design principles
+
+Unify client generation code into a single library.  We will be able to test
+this library individually on various Spark setups.  This will probably not
+usefully be automatic -- there is no automatic source for _new_ Spark
+setups, and it is not clear how often _existing_ Spark setups change.  But
+even being able to run a single command on a Spark cluster and get useful
+information will be very useful for investigation, helping customers probe
+their setup, and further development to support setups where we fail.
+
+This library will define an interface for _client acquisition_: given
```

Comment:
The client-acquisition interface can be defined in general terms so that in the future we'll be able to use it with other cloud providers if necessary (which will probably be).